kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt

```diff
@@ -0,0 +1,177 @@
+                    json.decodeFromJsonElement(serializer, serializedObject.data as JsonElement)
+
+                else ->
+                    throw org.axonframework.serialization.SerializationException("Cannot deserialize from content type ${serializedObject.contentType} to type ${serializedObject.type}. String and JsonElement are supported.")
+            }
+        } catch (ex: SerializationException) {
+            throw org.axonframework.serialization.SerializationException(
+                "Could not deserialize from content type ${serializedObject?.contentType} to type ${serializedObject?.type}",
+                ex
+            )
+        }
+    }
+
+    private fun <S, T> SerializedObject<S>.serializer(): KSerializer<T> =
+        classForType(type).serializer() as KSerializer<T>
+
+    /**
+     * When a type is compiled by the Kotlin compiler extension, a companion object
+     * is created which contains a method `serializer()`. This method should be called
+     * to get the serializer of the class.
+     *
+     * In a 'normal' serialization environment, you would call the MyClass.serializer()
+     * method directly. Here we are in a generic setting, and need reflection to call
+     * the method.
+     *
+     * This method caches the reflection mapping from class to serializer for efficiency.
+     */
+    private fun <T> Class<T>.serializer(): KSerializer<T> =
+        serializerCache.computeIfAbsent(this) {
+            // Class<T>: T must be non-null
+            val kClass = (this as Class<Any>).kotlin
+
+            val companion = kClass.companionObject
+                ?: throw SerializationException("Class $this has no companion object. Did you mark it as @Serializable?")
+
+            val serializerMethod = companion.java.getMethod("serializer")
+                ?: throw SerializationException("Class $this has no serializer() method. Did you mark it as @Serializable?")
+
+            serializerMethod.invoke(kClass.companionObjectInstance) as KSerializer<*>
+        } as KSerializer<T>
+
+    override fun classForType(type: SerializedType): Class<*> =
+        if (SerializedType.emptyType() == type) {
+            Void.TYPE
+        } else {
+            try {
+                Class.forName(type.name)
+            } catch (e: ClassNotFoundException) {
+                UnknownSerializedType::class.java
+            }
+        }
+
+    override fun typeForClass(type: Class<*>): SerializedType =
+        SimpleSerializedType(type.name, revisionOf(type))
+
+    private fun revisionOf(type: Class<*>): String? =
+        revisionResolver.revisionOf(type)
+
+    override fun getConverter(): Converter =
+        converter
+
+}
+
+class KotlinSerializerConfiguration {
+    var revisionResolver: RevisionResolver = AnnotationRevisionResolver()
+    var converter: Converter = ChainingConverter()
+    var json: Json = Json
+}
+
+fun kotlinSerializer(init: KotlinSerializerConfiguration.() -> Unit = {}): KotlinSerializer {
```

Comment:
Missing Javadoc. Public functions exposed as API should have a description, especially a function representing a DSL